powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/callbacks/CallbackUrlEventService.java

```diff
@@ -0,0 +1,302 @@
+     */
+    @Transactional(propagation = Propagation.REQUIRES_NEW)
+    public void dispatchInstantCallbackUrlEvent(final CallbackUrlEvent callbackUrlEvent) {
+        callbackUrlEventRepository.findById(callbackUrlEvent.callbackUrlEventEntityId())
+                .ifPresentOrElse(
+                        this::dispatchEvent,
+                        () -> { throw new IllegalStateException("Callback Url Event cannot be dispatched, because it was not found in database: callbackUrlEventEntityId=" + callbackUrlEvent.callbackUrlEventEntityId()); }
+                );
+    }
+
+    /**
+     * Dispatch again Callback URL Events, that previously failed and are eligible for another attempt.
+     */
+    @Transactional
+    public void dispatchFailedCallbackUrlEvents() {
+        final PageRequest pageRequest = PageRequest.of(0, powerAuthCallbacksConfiguration.getFailedCallbackUrlEventsRetryLimit());
+        callbackUrlEventRepository.findScheduledForRetry(LocalDateTime.now(), pageRequest)
+                .forEach(this::dispatchEvent);
+    }
+
+    /**
+     * Dispatch Callback URL Events in pending state.
+     */
+    @Transactional
+    public void dispatchPendingCallbackUrlEvents() {
+        final PageRequest pageRequest = PageRequest.of(0, powerAuthCallbacksConfiguration.getPendingCallbackUrlEventsDispatchLimit());
+        callbackUrlEventRepository.findPending(pageRequest)
+                .forEach(this::dispatchEvent);
+    }
+
+    /**
+     * Delete Callback URL Events, that are past their retention period.
+     */
+    @Transactional
+    public void deleteCallbackUrlEventsAfterRetentionPeriod() {
+        callbackUrlEventRepository.deleteCompletedAfterRetentionPeriod(LocalDateTime.now());
+    }
+
+    /**
+     * Dispatch Callback URL Event.
+     * @param callbackUrlEventEntity Event to dispatch.
+     */
+    private void dispatchEvent(final CallbackUrlEventEntity callbackUrlEventEntity) {
+
+        if (!maxAttemptsPositive(callbackUrlEventEntity)) {
+            logger.info("Callback URL Event {} has not positive max number of attempts.", callbackUrlEventEntity.getId());
+            final CallbackUrlEvent callbackUrlEvent = CallbackUrlConvertor.convert(callbackUrlEventEntity);
+            callbackUrlEventResponseHandler.handleSuccess(callbackUrlEvent);
+            return;
+        }
+
+        if (!shouldBeProcessed(callbackUrlEventEntity)) {
+            logger.debug("Callback URL Event {} should not be processed, state={}", callbackUrlEventEntity.getId(), callbackUrlEventEntity.getStatus());
+            return;
+        }
+
+        callbackUrlEventEntity.setStatus(CallbackUrlEventStatus.PROCESSING);
+        callbackUrlEventEntity.setTimestampNextCall(null);
+        callbackUrlEventEntity.setTimestampLastCall(LocalDateTime.now());
+        callbackUrlEventEntity.setAttempts(callbackUrlEventEntity.getAttempts() + 1);
+        callbackUrlEventRepository.save(callbackUrlEventEntity);
```

Comment:
Can we use `callbackUrlEventEntity = callbackUrlEventRepository.save(callbackUrlEventEntity)` to avoid issues in the future?